exla/lib/exla/defn.ex

```diff
@@ -2939,4 +2892,58 @@ defmodule EXLA.Defn do
   defp to_mlir_logical(%Value{} = value) do
     to_type(value, {:pred, 8})
   end
+
+  defp new_mlir_function(module_and_name, arg_shapes, return_shape, output_tuple?) do
+    in_shape = exla_shape(arg_shapes)
+    out_shape = exla_shape(return_shape)
+
+    out_shape =
+      if output_tuple? do
+        [EXLA.Shape.make_tuple_shape(out_shape)]
+      else
+        out_shape
+      end
+
+    EXLA.Builder.new_mlir(module_and_name, in_shape, out_shape)
+  end
+
+  defp exla_shape(tensors) when is_list(tensors) do
+    Enum.flat_map(tensors, &exla_shape/1)
+  end
+
+  defp exla_shape(tensors) when is_tuple(tensors) do
```

Comment:
I am not the biggest fan of such APIs here, it means we usually don't know the shape of the input. Could we rely on only one or the other?

Reply:
What do you mean by "one or the other"?